### ISSUE TITLE
Reducing problem data tensor can be a bottleneck

### ISSUE THREAD

**Issue by akshayka** [cvxgrp/cvxpy#1160]
**Describe the bug**
The `reduce_problem_data_tensor` function in `canonInterface.py` can be very slow.

**To Reproduce**
```python3
import cvxpy as cp
import numpy as np

m, n = 100000, 1000
X = np.random.randn(m, n)                                                              
y = np.random.randn(n)                                      
beta = cp.Variable(m)                                         
prob = cp.Problem(cp.Minimize(cp.sum_squares(y - X.T @ beta) +  cp.norm(beta, 1)))
prob.solve()
```

**Output**
![image](https://user-images.githubusercontent.com/1994308/97473793-9e3a8480-1908-11eb-8d71-220f8d9edced.png)

**Version**
 - OS: Ubuntu 20.10
 - CVXPY Version: 1.1.7

**Additional context**
`reduce_problem_data_tensor` was introduced in https://github.com/cvxgrp/cvxpy/pull/1010, where it was found that removing the rows in the tensor that are identically zero could in some cases yield orders of magnitude speed ups when multiplying the tensor by the parameter vector. This function is implemented in Python and performs a for loop over the nnz in A, which is possibly the bottleneck when nnz is large.

One option is to move this function into C. A better option would be to construct a more efficient implementation of this function in Python

**Comment by akshayka** [comment]
@sbarratt and I have separately looked into this, and neither of us was able to find a Python implementation that was much faster (and didn't OOM our machines on large problems). So it seems like a C++ port of the current implementation would be the way to go (with say a pybind11 binding).

**Comment by Mojusko** [comment]
Same problem here? Is there a way to disable it?

**Comment by rileyjmurray** [comment]
I don't know anything about disabling this function and falling back on an old implementation (it seems unlikely that we'd leave something like that in). However, as a more general note, I think it might be possible to get some noticeable performance improvements to the current python implementation by making very small changes. I've had to do an almost identical operation before, just with zero columns of a matrix

https://github.com/rileyjmurray/sageopt/blob/001fc80b24fe5ee7df24890af6d4d935ffd2983a/sageopt/coniclifts/utilities.py#L59-L66

The difference is that I iterate over the lists/arrays directly, rather than iterating over an index that I use to access arrays. So for example, I'd change
```
    reduced_row = np.zeros(old_row.shape)
    for i in range(reduced_row.size):
        reduced_row[i] = old_row_to_new_row[old_row[i]]
```
to
```
    reduced_row = np.array([old_row_to_new_row[orow] for orow in old_row])
```

**Comment by Mojusko** [comment]
Thank you. This actually works much better. I am very surprised, we should create a pull request.

**Comment by akshayka** [comment]
Yes, please do!

**Comment by rileyjmurray** [comment]
This is resolved by #1235. If someone has a new test case showing that the change in #1235 leaves significant room for improvement, we can re-open this issue and discuss it.
